### PR TITLE
Disallow Insert with Duplicate key update and Replace Into queries on foreign key column, set locks on fk queries

### PR DESCRIPTION
**Pull request by harshit-gangal**
## Description

Currently, the planner does not support the below cases. Therefore rejecting it for now.

1. Insert query with `on duplicate key update` when the update list contains columns with foreign keys.
2. Non-literal value on FK columns in update query (earlier was on any column)
3. Replace query

This PR also adds locks on select queries created for FK support
 - FK Verify select query will take `lock in share mode` lock
 - FK Cascade selection query will take `for update` lock

## Related Issue(s)

- #12967 

## Checklist

-   [X] Tests were added or are not required
-   [X] Did the new or modified tests pass consistently locally and on the CI

### CODE REVIEW

File: go/vt/vtgate/planbuilder/operators/delete.go [hunk 1]
```diff
@@ -69,3 +73,186 @@ func (d *Delete) ShortDescription() string {
...
+	}
+	// If the delete statement has a limit, we don't support it yet.
+	if deleteStmt.Limit != nil {
+		return nil, vterrors.VT12001("foreign keys management at vitess with limit")
+	}
+
+	return createFkCascadeOpForDelete(ctx, delOp, delClone, childFks)
+}
+
+func createDeleteOperator(ctx *plancontext.PlanningContext, deleteStmt *sqlparser.Delete, qt *QueryTable, vindexTable *vindexes.Table, routing Routing) (ops.Operator, error) {
```
Comment (systay) [inline go/vt/vtgate/planbuilder/operators/delete.go:119]: nitpick: this line is 176 characters long. IMHO, that is too many

**Review by GuptaManan100 — APPROVED** [review]

File: go/vt/vtgate/planbuilder/insert.go [hunk 2]
```diff
@@ -45,8 +45,11 @@ func gen4InsertStmtPlanner(version querypb.ExecuteOptions_PlannerVersion, insStm
 	// Check single unsharded. Even if the table is for single unsharded but sequence table is used.
 	// We cannot shortcut here as sequence column needs additional planning.
 	ks, tables := ctx.SemTable.SingleUnshardedKeyspace()
-	if ks != nil && tables[0].AutoIncrement == nil {
-		if fkManagementNotRequiredForInsert(ctx, tables[0], sqlparser.UpdateExprs(insStmt.OnDup), insStmt.Action == sqlparser.ReplaceAct) {
+	FkPlanNeeded := false
```
Comment (systay) [inline go/vt/vtgate/planbuilder/insert.go:48]: nitpick: why capitalized?

File: go/vt/vtgate/planbuilder/operator_transformers.go [hunk 3]
```diff
@@ -439,6 +439,12 @@ func transformRoutePlan(ctx *plancontext.PlanningContext, op *operators.Route) (
 
 	switch stmt := stmt.(type) {
 	case sqlparser.SelectStatement:
+		if op.Lock != sqlparser.NoLock {
+			stmt.SetLock(op.Lock)
+		}
+		if op.Comments != nil {
```
Comment (systay) [inline go/vt/vtgate/planbuilder/operator_transformers.go:445]: comments apply to all statement types and not just to SELECT and UNION, right?

File: go/vt/vtgate/planbuilder/operators/operator.go [hunk 4]
```diff
@@ -89,6 +89,21 @@ func PlanQuery(ctx *plancontext.PlanningContext, stmt sqlparser.Statement) (ops.
 		return nil, ctx.SemTable.NotSingleRouteErr
 	}
 
+	// set lock and comments on the route to be set on the sql query on conversion.
```
Comment (systay) [inline go/vt/vtgate/planbuilder/operators/operator.go:92]: this should be extracted into it's own method to follow the rest of `PlanQuery` flow

**Review by systay — APPROVED** [review]

File: go/vt/vtgate/planbuilder/operator_transformers.go (hunk 3 above)
Comment (harshit-gangal) [reply go/vt/vtgate/planbuilder/operator_transformers.go:445]: yes, I only removed setMiscFunc from the select plan, that's why it is used only here. Let me check for other statement types and if I can remove them from these places as well.

Comment (harshit-gangal) [reply go/vt/vtgate/planbuilder/operator_transformers.go:445]: In DML operator transform, the original AST is provided which contains the comment
```go
func buildDML(op OpWithAST, qb *queryBuilder) {
	qb.stmt = op.Statement()
	qb.dmlOperator = op
}
```

**Pull request merged**